File: backend/app/services/pncp_service.py
"""
Serviço para integração com API do PNCP (Portal Nacional de Contratações Públicas)
"""
import httpx
from typing import List, Optional, Dict, Any
from datetime import datetime, timedelta
import logging

logger = logging.getLogger(__name__)


class PNCPService:
    """
    Serviço para integração com API pública do PNCP
    Documentação: https://pncp.gov.br/api/pncp/swagger-ui/index.html
    """

    BASE_URL = "https://pncp.gov.br/api/consulta/v1"

    def __init__(self):
        self.client = httpx.AsyncClient(
            base_url=self.BASE_URL,
            timeout=30.0,
            limits=httpx.Limits(
                max_keepalive_connections=5,
                max_connections=10
            )
        )

    async def buscar_contratos(
        self,
        data_inicial: str,
        data_final: str,
        pagina: int = 1,
        tamanho_pagina: int = 500,
        cnpj_orgao: Optional[str] = None,
        uf: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Busca contratos no PNCP

        Args:
            data_inicial: Data inicial no formato YYYYMMDD (ex: 20250101)
            data_final: Data final no formato YYYYMMDD
            pagina: Número da página (padrão: 1)
            tamanho_pagina: Quantidade de registros por página (10-500)
            cnpj_orgao: CNPJ do órgão (opcional)
            uf: Sigla da UF (opcional)

        Returns:
            Dict com dados de contratos e paginação
        """
        try:
            params = {
                "dataInicial": data_inicial,
                "dataFinal": data_final,
                "pagina": pagina,
                "tamanhoPagina": min(tamanho_pagina, 500)  # Máximo 500
            }

            if cnpj_orgao:
                params["cnpjOrgao"] = cnpj_orgao
            if uf:
                params["uf"] = uf

            logger.info(f"Buscando contratos PNCP: {params}")

            response = await self.client.get("/contratos", params=params)
            response.raise_for_status()

            data = response.json()
            return data

        except httpx.HTTPStatusError as e:
            logger.error(f"Erro HTTP ao buscar contratos PNCP: {e.response.status_code} - {e.response.text}")
            raise
        except Exception as e:
            logger.error(f"Erro ao buscar contratos PNCP: {e}")
            raise

    async def buscar_contrato_detalhes(
        self,
        cnpj: str,
        ano: int,
        sequencial: int
    ) -> Dict[str, Any]:
        """
        Busca detalhes de um contrato específico

        Args:
            cnpj: CNPJ do órgão
            ano: Ano do contrato
            sequencial: Número sequencial do contrato

        Returns:
            Dict com detalhes do contrato
        """
        try:
            url = f"/contratos/{cnpj}/{ano}/{sequencial}"
            logger.info(f"Buscando detalhes do contrato: {url}")

            response = await self.client.get(url)
            response.raise_for_status()

            return response.json()

        except httpx.HTTPStatusError as e:
            logger.error(f"Erro ao buscar detalhes do contrato: {e.response.status_code}")
            raise
        except Exception as e:
            logger.error(f"Erro ao buscar detalhes do contrato: {e}")
            raise

    async def buscar_itens_contrato(
        self,
        cnpj: str,
        ano: int,
        sequencial: int
    ) -> List[Dict[str, Any]]:
        """
        Busca itens de um contrato específico

        Args:
            cnpj: CNPJ do órgão
            ano: Ano do contrato
            sequencial: Número sequencial do contrato

        Returns:
            Lista de itens do contrato
        """
        try:
            url = f"/contratos/{cnpj}/{ano}/{sequencial}/itens"
            logger.info(f"Buscando itens do contrato: {url}")

            response = await self.client.get(url)
            response.raise_for_status()

            data = response.json()
            return data.get("data", [])

        except httpx.HTTPStatusError as e:
            if e.response.status_code == 404:
                logger.warning(f"Itens não encontrados para contrato {cnpj}/{ano}/{sequencial}")
                return []
            logger.error(f"Erro ao buscar itens: {e.response.status_code}")
            raise
        except Exception as e:
            logger.error(f"Erro ao buscar itens: {e}")
            raise

    async def buscar_compras(
        self,
        data_inicial: str,
        data_final: str,
        pagina: int = 1,
        tamanho_pagina: int = 500,
        modalidade: Optional[str] = None,
        uf: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Busca processos de compra/licitações no PNCP

        Args:
            data_inicial: Data inicial no formato YYYYMMDD
            data_final: Data final no formato YYYYMMDD
            pagina: Número da página
            tamanho_pagina: Quantidade de registros (10-500)
            modalidade: Código da modalidade (opcional)
            uf: Sigla da UF (opcional)

        Returns:
            Dict com dados de compras e paginação
        """
        try:
            params = {
                "dataInicial": data_inicial,
                "dataFinal": data_final,
                "pagina": pagina,
                "tamanhoPagina": min(tamanho_pagina, 500)
            }

            if modalidade:
                params["codigoModalidade"] = modalidade
            if uf:
                params["uf"] = uf

            logger.info(f"Buscando compras PNCP: {params}")

            response = await self.client.get("/compras", params=params)
            response.raise_for_status()

            return response.json()

        except httpx.HTTPStatusError as e:
            logger.error(f"Erro ao buscar compras: {e.response.status_code} - {e.response.text}")
            raise
        except Exception as e:
            logger.error(f"Erro ao buscar compras: {e}")
            raise

    async def buscar_compra_detalhes(
        self,
        cnpj: str,
        ano: int,
        sequencial: int
    ) -> Dict[str, Any]:
        """
        Busca detalhes de uma compra específica

        Args:
            cnpj: CNPJ do órgão
            ano: Ano da compra
            sequencial: Número sequencial da compra

        Returns:
            Dict com detalhes da compra
        """
        try:
            url = f"/compras/{cnpj}/{ano}/{sequencial}"
            logger.info(f"Buscando detalhes da compra: {url}")

            response = await self.client.get(url)
            response.raise_for_status()

            return response.json()

        except httpx.HTTPStatusError as e:
            logger.error(f"Erro ao buscar detalhes da compra: {e.response.status_code}")
            raise
        except Exception as e:
            logger.error(f"Erro ao buscar detalhes da compra: {e}")
            raise

    def formatar_data_pncp(self, data: datetime) -> str:
        """
        Formata datetime para o formato aceito pelo PNCP (YYYYMMDD)

        Args:
            data: Objeto datetime

        Returns:
            String no formato YYYYMMDD
        """
        return data.strftime("%Y%m%d")

    def obter_periodo_ultima_semana(self) -> tuple[str, str]:
        """
        Retorna datas (inicial e final) da última semana no formato PNCP

        Returns:
            Tupla (data_inicial, data_final) em formato YYYYMMDD
        """
        hoje = datetime.now()
        semana_atras = hoje - timedelta(days=7)
        return (
            self.formatar_data_pncp(semana_atras),
            self.formatar_data_pncp(hoje)
        )

    def obter_periodo_ultimo_mes(self) -> tuple[str, str]:
        """
        Retorna datas (inicial e final) do último mês no formato PNCP

        Returns:
            Tupla (data_inicial, data_final) em formato YYYYMMDD
        """
        hoje = datetime.now()
        mes_atras = hoje - timedelta(days=30)
        return (
            self.formatar_data_pncp(mes_atras),
            self.formatar_data_pncp(hoje)
        )

    async def close(self):
        """Fecha a conexão HTTP"""
        await self.client.aclose()

    async def __aenter__(self):
        """Context manager async enter"""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Context manager async exit"""
        await self.close()

File: backend/src/Services/PNCPService.php
<?php

namespace App\Services;

use App\Models\Licitacao;
use App\Models\Orgao;
use App\Repositories\LicitacaoRepository;
use App\Repositories\OrgaoRepository;
use Exception;

/**
 * Service para integração com a API do PNCP
 *
 * Documentação oficial: https://pncp.gov.br/api/consulta/swagger-ui/index.html
 */
class PNCPService
{
    private const BASE_URL = 'https://pncp.gov.br/api/consulta/v1';
    private const TIMEOUT = 30; // segundos
    private const MAX_RETRIES = 3;

    private LicitacaoRepository $licitacaoRepo;
    private OrgaoRepository $orgaoRepo;
    private array $stats = [
        'novos' => 0,
        'atualizados' => 0,
        'erros' => 0,
        'pulados' => 0,
    ];

    public function __construct()
    {
        $this->licitacaoRepo = new LicitacaoRepository();
        $this->orgaoRepo = new OrgaoRepository();
    }

    /**
     * Sincronizar licitações do PNCP
     *
     * @param array $filtros Filtros para busca
     * @return array Estatísticas da sincronização
     */
    public function sincronizarLicitacoes(array $filtros = []): array
    {
        $this->resetStats();
        $iniciado = date('Y-m-d H:i:s');

        try {
            // Parâmetros padrão
            $params = array_merge([
                'dataInicial' => date('Ymd', strtotime('-7 days')), // Últimos 7 dias
                'dataFinal' => date('Ymd'),
                'tamanhoPagina' => 50,
                'pagina' => 1,
            ], $filtros);

            $totalPaginas = 1;
            $paginaAtual = $params['pagina'];

            do {
                echo "Sincronizando página {$paginaAtual} de {$totalPaginas}...\n";

                $params['pagina'] = $paginaAtual;
                $response = $this->fazerRequisicao('/contratos', $params);

                if (!$response || empty($response['data'])) {
                    echo "Nenhum dado retornado na página {$paginaAtual}\n";
                    break;
                }

                // Atualizar total de páginas
                if (isset($response['pagination'])) {
                    $totalPaginas = $response['pagination']['totalPages'] ?? 1;
                }

                // Processar cada licitação
                foreach ($response['data'] as $item) {
                    $this->processarLicitacao($item);
                }

                $paginaAtual++;

                // Pausa entre requisições para não sobrecarregar a API
                usleep(500000); // 0.5 segundo

            } while ($paginaAtual <= $totalPaginas && $paginaAtual <= 10); // Limite de 10 páginas por execução

            $finalizado = date('Y-m-d H:i:s');

            return [
                'sucesso' => true,
                'stats' => $this->stats,
                'iniciado' => $iniciado,
                'finalizado' => $finalizado,
                'duracao' => strtotime($finalizado) - strtotime($iniciado),
            ];

        } catch (Exception $e) {
            return [
                'sucesso' => false,
                'erro' => $e->getMessage(),
                'stats' => $this->stats,
                'iniciado' => $iniciado,
                'finalizado' => date('Y-m-d H:i:s'),
            ];
        }
    }

    /**
     * Processar uma licitação individual
     */
    private function processarLicitacao(array $item): void
    {
        try {
            // Verificar se já existe
            $pncpId = $item['numeroControlePNCP'] ?? $item['numeroControlePncpCompra'] ?? $item['numeroCompra'] ?? $item['id'] ?? null;

            if (!$pncpId) {
                $this->stats['erros']++;
                echo "  ✗ Licitação sem ID\n";
                return;
            }

            $licitacaoExistente = $this->licitacaoRepo->findByPncpId($pncpId);

            // Garantir que o órgão existe
            $orgaoId = $item['unidadeOrgao']['codigoUnidade'] ?? $item['codigoUnidadeCompradora'] ?? $item['orgaoId'] ?? null;

            if ($orgaoId) {
                $this->garantirOrgao($orgaoId, $item);
            }

            // Criar/atualizar licitação
            $licitacao = $this->mapearLicitacaoDoPNCP($item);

            if ($licitacaoExistente) {
                $licitacao->id = $licitacaoExistente->id;
                $this->licitacaoRepo->update($licitacao);
                $this->stats['atualizados']++;
                echo "  ↻ Atualizada: {$pncpId}\n";
            } else {
                $this->licitacaoRepo->create($licitacao);
                $this->stats['novos']++;
                echo "  ✓ Nova: {$pncpId}\n";
            }

        } catch (Exception $e) {
            $this->stats['erros']++;
            echo "  ✗ Erro: " . $e->getMessage() . "\n";
        }
    }

    /**
     * Garantir que o órgão existe no banco
     */
    private function garantirOrgao(string $orgaoId, array $dadosLicitacao): void
    {
        $orgaoExistente = $this->orgaoRepo->findById($orgaoId);

        if (!$orgaoExistente) {
            // Criar órgão básico com dados da licitação
            $orgao = new Orgao();
            $orgao->id = $orgaoId;
            $orgao->cnpj = $dadosLicitacao['cnpjUnidadeCompradora'] ?? str_pad($orgaoId, 14, '0', STR_PAD_LEFT);
            $orgao->razao_social = $dadosLicitacao['nomeUnidadeCompradora'] ?? 'Órgão ' . $orgaoId;
            $orgao->nome_fantasia = $dadosLicitacao['siglaUnidadeCompradora'] ?? null;
            $orgao->esfera = 'MUNICIPAL'; // Assumir municipal por padrão
            $orgao->poder = 'EXECUTIVO';
            $orgao->uf = $dadosLicitacao['uf'] ?? 'SP';
            $orgao->municipio = $dadosLicitacao['municipio'] ?? null;

            $this->orgaoRepo->upsert($orgao);
        }
    }

    /**
     * Mapear dados do PNCP para modelo Licitacao
     */
    private function mapearLicitacaoDoPNCP(array $item): Licitacao
    {
        $licitacao = new Licitacao();

        $licitacao->pncp_id = $item['numeroControlePNCP'] ?? $item['numeroControlePncpCompra'] ?? $item['numeroCompra'] ?? $item['id'];
        $licitacao->orgao_id = $item['unidadeOrgao']['codigoUnidade'] ?? $item['codigoUnidadeCompradora'] ?? $item['orgaoId'];
        $licitacao->numero = $item['numeroContratoEmpenho'] ?? $item['numeroProcesso'] ?? $item['numero'] ?? $licitacao->pncp_id;
        $licitacao->objeto = $item['objetoContrato'] ?? $item['objeto'] ?? $item['descricao'] ?? '';
        $licitacao->modalidade = $this->mapearModalidade($item['tipoContrato']['id'] ?? $item['codigoModalidade'] ?? $item['modalidade'] ?? 0);
        $licitacao->situacao = 'ATIVO'; // Contratos geralmente estão ativos
        $licitacao->valor_estimado = $item['valorGlobal'] ?? $item['valorInicial'] ?? $item['valorEstimado'] ?? $item['valor'] ?? null;

        // Datas
        $licitacao->data_publicacao = $this->formatarData($item['dataPublicacaoPncp'] ?? $item['dataPublicacao'] ?? date('Y-m-d'));
        $licitacao->data_abertura = $this->formatarData($item['dataAssinatura'] ?? $item['dataAberturaProposta'] ?? $item['dataAbertura'] ?? null);
        $licitacao->data_encerramento = $this->formatarData($item['dataVigenciaFim'] ?? $item['dataEncerramentoProposta'] ?? $item->dataEncerramento ?? null);

        // Localização
        $licitacao->uf = strtoupper($item['unidadeOrgao']['ufSigla'] ?? $item['uf'] ?? $item['unidadeFederacao'] ?? 'SP');
        $licitacao->municipio = $item['unidadeOrgao']['municipioNome'] ?? $item['municipio'] ?? $item['nomeMunicipio'] ?? '';

        // URLs
        $licitacao->url_edital = $item['urlCipi'] ?? $item['urlEdital'] ?? null;
        $licitacao->url_pncp = $item['linkSistemaOrigem'] ?? "https://pncp.gov.br/app/contratos/{$licitacao->pncp_id}";

        // Órgão
        $licitacao->nome_orgao = $item['unidadeOrgao']['nomeUnidade'] ?? $item['orgaoEntidade']['razaoSocial'] ?? $item['nomeUnidadeCompradora'] ?? $item['nomeOrgao'] ?? '';
        $licitacao->cnpj_orgao = $item['orgaoEntidade']['cnpj'] ?? $item['cnpjUnidadeCompradora'] ?? $item['cnpjOrgao'] ?? '';

        return $licitacao;
    }

    /**
     * Mapear código de modalidade do PNCP
     */
    private function mapearModalidade(int $codigo): string
    {
        $modalidades = [
            1 => 'CONCORRENCIA',
            2 => 'TOMADA_PRECOS',
            3 => 'CONVITE',
            4 => 'CONCURSO',
            5 => 'LEILAO',
            6 => 'PREGAO_ELETRONICO',
            7 => 'PREGAO_PRESENCIAL',
            8 => 'DISPENSA',
            9 => 'INEXIGIBILIDADE',
            10 => 'DIALOGO_COMPETITIVO',
            11 => 'CREDENCIAMENTO',
            12 => 'PRE_QUALIFICACAO',
        ];

        return $modalidades[$codigo] ?? 'OUTROS';
    }

    /**
     * Mapear situação da licitação
     */
    private function mapearSituacao(string $situacao): string
    {
        $situacao = strtoupper($situacao);

        $mapeamento = [
            'ATIVA' => 'ATIVO',
            'PUBLICADA' => 'ATIVO',
            'EM_ANDAMENTO' => 'ATIVO',
            'ENCERRADA' => 'ENCERRADO',
            'CANCELADA' => 'CANCELADO',
            'SUSPENSA' => 'SUSPENSO',
            'REVOGADA' => 'CANCELADO',
        ];

        return $mapeamento[$situacao] ?? 'ATIVO';
    }

    /**
     * Formatar data do PNCP
     */
    private function formatarData(?string $data): ?string
    {
        if (!$data) {
            return null;
        }

        // Se já está no formato Y-m-d H:i:s
        if (preg_match('/^\d{4}-\d{2}-\d{2}/', $data)) {
            return $data;
        }

        // Se está no formato Ymd (20250101)
        if (preg_match('/^(\d{4})(\d{2})(\d{2})$/', $data, $matches)) {
            return "{$matches[1]}-{$matches[2]}-{$matches[3]}";
        }

        // Tentar outros formatos
        try {
            $date = new \DateTime($data);
            return $date->format('Y-m-d H:i:s');
        } catch (Exception $e) {
            return null;
        }
    }

    /**
     * Fazer requisição à API do PNCP
     */
    private function fazerRequisicao(string $endpoint, array $params = [], int $tentativa = 1): ?array
    {
        $url = self::BASE_URL . $endpoint . '?' . http_build_query($params);

        $curl = curl_init();

        curl_setopt_array($curl, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => self::TIMEOUT,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTPHEADER => [
                'Accept: application/json',
                'User-Agent: Licita.pub/1.0',
            ],
        ]);

        $response = curl_exec($curl);
        $httpCode = curl_getinfo($curl, CURLINFO_HTTP_CODE);
        $error = curl_error($curl);

        curl_close($curl);

        // Retry em caso de erro
        if ($httpCode !== 200 || !$response) {
            if ($tentativa < self::MAX_RETRIES) {
                echo "  Erro HTTP {$httpCode}, tentando novamente ({$tentativa}/" . self::MAX_RETRIES . ")...\n";
                sleep(2 * $tentativa); // Backoff exponencial
                return $this->fazerRequisicao($endpoint, $params, $tentativa + 1);
            }

            throw new Exception("Erro na requisição: HTTP {$httpCode} - {$error}");
        }

        $data = json_decode($response, true);

        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new Exception('Erro ao decodificar JSON: ' . json_last_error_msg());
        }

        return $data;
    }

    /**
     * Resetar estatísticas
     */
    private function resetStats(): void
    {
        $this->stats = [
            'novos' => 0,
            'atualizados' => 0,
            'erros' => 0,
            'pulados' => 0,
        ];
    }

    /**
     * Obter estatísticas da última sincronização
     */
    public function getStats(): array
    {
        return $this->stats;
    }
}
